\documentclass[11pt]{article}
\usepackage{setspace}
\usepackage[letterpaper, margin=0.75in]{geometry}
\usepackage{amsmath}

%\usepackage[sort&compress]{natbib}
\begin{document}
  \title{Description of the mixture model implemented in the R package ribModel}
  \author{Cedric Landerer} % Your name
  \date{\today}
  \maketitle
  \doublespacing

  
  \section{General}
  \begin{table}[h]
    \centering
    \caption{Variable legend}
    \label{var_desc}
    \begin{tabular}{|l|l|}
      \cline{1-2}
      \textbf{Variable} & \textbf{Description}		\\ \cline{1-2}
      $G$		& Number of genes		\\ \cline{1-2}
      $g$		& Gene index			\\ \cline{1-2}
      $A$		& Number of Amino Acids		\\ \cline{1-2}
      $a$		& Amino acid index		\\ \cline{1-2}
      $C$		& Number of Codons		\\ \cline{1-2}
      $c$		& Codon index			\\ \cline{1-2}
      $\vec{\zeta}$	& Vector of codon counts	\\ \cline{1-2}
      $\zeta_c$		& Codon count for codon $c$	\\ \cline{1-2}
      $f$		& log Likelihood function	\\ \cline{1-2}
      $\Theta$		& Set of parameters for $f$	\\ \cline{1-2}
      $I$		& Number of $\phi$ obs.		\\ \cline{1-2}
      $\iota$		& $\phi$ obs. index		\\ \cline{1-2}
      $\Phi$		& Obs. $\phi$ value		\\ \cline{1-2}
      $\phi$		& estm. $\phi$ value		\\ \cline{1-2}
      $P$		& Posterior			\\ \cline{1-2}
      $\Upsilon$	& Posterior Jacobian adjusted	\\ \cline{1-2}
      $z$		& Mixture			\\ \cline{1-2}
      $Z$		& Number of mixtures		\\ \cline{1-2}
    \end{tabular}
  \end{table}
  
  
  \begin{table}[h]
    \centering
    \caption{Parameter distributions}
    \label{param_dist}
    \begin{tabular}{|l|l|l|}
      \cline{1-3}
      \textbf{Parameter} & \textbf{Distribution}				& \textbf{proposal}					\\ \cline{1-3}
      $\phi$		& $LN(-\frac{s_{\phi}^2}{2}, s_{\phi})$	& $\log(\phi') \sim N(\log(\phi), \sigma_{\phi})$		\\ \cline{1-3}
      $s_{\phi}$	& $U(0, \infty)$	& $\log(s_\phi') \sim N(\log(s_\phi), \sigma_{s_{\phi}})$		\\ \cline{1-3}
      $\Delta M$	& $N(0, \sigma)$	& $\Delta M' \sim N(\Delta M, \sigma)$		\\ \cline{1-3}
      $\Delta \eta$	& $U(-\infty, \infty)$	& $\Delta \eta' \sim N(\Delta \eta, \sigma)$		\\ \cline{1-3}
      $p_z$		& $Dir(\alpha)$		& Gibbs sampled	\\ \cline{1-3}
      $A_{\phi}$	& $U(-\infty, \infty)$	& $A_\phi' \sim N(A_\phi, \sigma_{A_{\phi}})$		\\ \cline{1-3}
      $s_{\epsilon}$	& $G^{-1}(\frac{G-1}{2}, \frac{S^2}{2})$	& Gibbs sampled	\\ \cline{1-3}
    \end{tabular}
  \end{table}
  
  \subsection{Calculating the Posterior trace}
  We calculate the unscaled log posterior as shown in equation \ref{post_trace} as
  \begin{equation}
    P(\phi | \zeta, z, \Delta M, \Delta \eta, s_{\phi}, A_{\phi}, s_{\epsilon}) \propto \sum_g^G \sum_z^Z \left(p_{z,g}P_g(\phi_{z,g}|\vec{\zeta}, \Theta_z)\right) + \sum_a^A \log(\pi(\Delta M_a| 0, \sigma))
    \label{post_trace}
  \end{equation}
  where $p_{z,g}$ is the probability of gene $g$ being assigned to mixture $z$ (see eqn \ref{mix_assign_prob}), 
  and $\Upsilon_g(\phi_{z,g}|\zeta, \Theta_z)$ is the log posterior for gene $g$ under the parameters defining mixture $z$ (see eqn \ref{roc_posterior_jacobian}). We assume $\Delta M$ to be normal distributed with $\pi(\Delta M_a| 0, \sigma)$, 
  ($\sigma = x \hspace{1mm} \forall a$), and take it as prior into account when calculating the unscaled log posterior. 
  
  \subsection{Calculate mixture assignment}
  Genes are assigned to mixtures based on there likelihood calculated under each mixture weighted by the probability of that mixture.
  The probability for a gene being in each mixture is calculated as shown in eqn. \ref{mix_assign_prob} and used to draw the mixture assignment from a multinomial distribution $z_g \sim \text{Multinomial}(Z, p_{1,g} \ldots p_{Z,g})$  
  \begin{equation}
    p_{z,g} = \exp \left(\frac{\log(p_z) + P_g(\phi_{z,g}|\vec{\zeta}, \Theta_z)}{\sum_z^Z \log(p_z) + P_g(\phi_{z,g}|\vec{\zeta}, \Theta_z)}\right)
    \label{mix_assign_prob}
  \end{equation}

  
  \section{ROC}
  
  We calculate the probability for each synonymous codon coding for an amino acid using equation \ref{codon_prob}.
  \begin{equation}
    p_{g,c} = \frac{-\Delta M_c - \Delta\eta_c \phi_g}{\sum_c^C -\Delta M_c - \Delta\eta_c \phi_g}
    \label{codon_prob}
  \end{equation}
  where $\Delta M$ is the difference in mutation bias between the current codon and a chosen reference codon, and $\Delta \eta$ is the difference in
  translation efficiency. 
  
  We assume codon count being distributed according to a multinomial distribution where the probability of each synonymous codon is given by equation \ref{codon_prob}.
  The log likelihood for a set of parameters describing each amino acid is therefore given by equation \ref{roc_lik_aa}
  \begin{equation}
    f_a(\cdot|\zeta, \Theta_z) = \sum_c^C \log(p_{g,c}) \zeta_c
    \label{roc_lik_aa}
  \end{equation}
  

  \subsection{Accept/Reject $\phi$}
  To accept/reject a proposed $\phi$ value, we calculate the log posterior as follows.
  The likelihood function is the sum over all likelihoods for the amino acids (eqn \ref{roc_lik_aa}) in the gene of interest shown in equation \ref{roc_lik_gene}.
  \begin{equation}
    f_g(\phi_g|\zeta, \Theta_z) = \sum_a^A f_a(\phi_g|\vec{\zeta}, \Theta_z)
    \label{roc_lik_gene}
  \end{equation}
  A log normal prior ($p(\phi_g|-\frac{s_{\phi}^2}{2}, s_{\phi})$) on $\phi$ is assumed.
  \begin{equation}
    P_g(\phi_g|\zeta, \Theta_z) = f_g(\phi_g|\zeta, \Theta_z) + \log(p(\phi_g|-\frac{s_{\phi}^2}{2}, s_{\phi})) + \sum_{\iota}^I \log(p(\log{\Phi_{g,\iota}}| \log{\phi_g} + {A_{\phi}}_{\iota}, {s_{\epsilon}}_{\iota}))
    \label{roc_posterior}
  \end{equation}
  All terms $\log(p(\log{\Phi_{g,\iota}} + {A_{\phi}}_{\iota}| \log{\phi_g}, {s_{\epsilon}}_{\iota}))$ are $0$ if no observed $\phi$ values $\Phi$ are available. 
  
  Since we assume the $\phi$ values to be log normal distributed, we take the reverse jumb probability into account in equation \ref{roc_posterior_jacobian}.
  \begin{equation}
    \Upsilon_g(\phi_g|\zeta, \Theta_z) = P_g(\phi_g|\zeta, \Theta_z) - \log(\phi_g)
    \label{roc_posterior_jacobian}
  \end{equation}
  The ratio of two log normal distributions can be simplified to the ratio $\frac{\phi}{\phi'}$ which results in the term $\log{\phi}$ on the log scale.
  
  The acceptance/rejection of a proposed $\phi_g'$ is performed on the logscale and described in equation \ref{a/r} where $r \sim Exp(1)$ and $\alpha$ is given by equation \ref{log_acceptance_ratio}
  \begin{equation}
    \phi_g = 
    \begin{cases}
      \phi_g',	& \text{if } -r < \alpha \\
      \phi_g,	& \text{else } \\
    \end{cases}
     \label{a/r}
  \end{equation}  
  \begin{equation}
    \alpha = \Upsilon_g(\phi_g'|\zeta, \Theta_z) - \Upsilon_g(\phi_g|\zeta, \Theta_z)
    \label{log_acceptance_ratio}
  \end{equation}
  
  \subsection{Accept/Reject $s_{\phi}$}
  Since we are not interested in the likelihood for $s_{\phi}$, we calculate the likelihood ratio necessary for the acceptance/rejection step directly.
  Since $s_{\phi}$ has to be positive, we propose a new $\log{s_{\phi}}$ using a normal distributed random walk.
  Therefore it follows that $s_{\phi}$ is drawn from a log-normal distribution. This forces us to take the reverse jump probability into account since the log-normal distribution is not symmetric. The ratio can be simplified since symmetric
  elements will cancel. We take the reverse jumb probability ratio for each mixture element into accountas as shown in equation \ref{jacobian_ratio_sphi}.
  
  \begin{equation}
    \Delta J = \frac{\log(J)}{\log(J')} = \sum_z^Z -(\log({s_{\phi}}_z) - \log({s_{\phi}'}_z))
    \label{jacobian_ratio_sphi}
  \end{equation}
  We then calculate the probability ratio of the data (here the current $\phi$ values estimated in the previous step) given the current and proposed $s_{\phi}$ values as shown in equation \ref{loglik_ratio_sphi}.
  \begin{equation}
    \Delta f(s_{\phi}|\phi_z) = \sum_g^G \left(\log(LN(\phi_{g,z} | -\frac{{s_{\phi}}_z^2}{2}, {s_{\phi}}_z)) - \log(LN(\phi_{g,z} | -\frac{{s_{\phi}'}_z^2}{2}, {s_{\phi}'}_z)) \right)
    \label{loglik_ratio_sphi}
  \end{equation}
  The reverse jump probability ratio and the likelihood ratio are combined on the log scale to obtain the log acceptance ratio $\alpha$ as shown in equation \ref{alpha_sphi}.
  \begin{equation}
    \alpha = \Delta f_z(s_{\phi}|\phi_z) + \Delta J
    \label{alpha_sphi}
  \end{equation}
  
 
  The acceptance/rejection of proposed $s_{\phi}'$ is given by equation \ref{a/r_sphi} where $r \sim Exp(1)$ and $\alpha$ is given by equation \ref{alpha_sphi}.
  \begin{equation}
    \phi_g = 
    \begin{cases}
      s_{\phi}',	& \text{if } -r < \alpha \\
      s_{\phi},		& \text{else } \\
    \end{cases}
     \label{a/r_sphi}
  \end{equation}    
  
  \subsection{Accept/Reject $\Delta M$ and $\Delta \eta$}
  
  The codon specific parameter $\Delta M$ and $\Delta \eta$ are acceptted/rejected together for each amino acid.
  A new set of $\Delta M'$ and $\Delta \eta'$ values $\theta$, is proposed from a normal distribution $\theta' \sim N(\theta, \Sigma)$, where $\Sigma$ is a covariance matrix where only the variance terms on the 
  diagonal are non zero. We assume as prior term a normal distribution form $\Delta M$ and a uniform prior for $\Delta \eta$.
  The posterior is for a set of $\Delta M$ and $\Delta \eta$, $\theta$ is calculated as
    \begin{equation}
    P_a(\Delta M_a, \Delta \eta_a|\zeta, \phi, z) = \sum_g^G f_a(\Delta M_a, \Delta \eta_a|\zeta, \phi_{g,z}) + \log(\pi(\Delta M_a| 0, \sigma))
    \label{roc_lik_csp}
  \end{equation}
  where $\pi(\Delta M_a| 0, \sigma)$ is the normal distributed prior on $\Delta M$.
  
  The acceptance/rejection of a proposed set of $\Delta M_a$ and $\Delta \eta_A$ is given by equation \ref{a/r_aa} where $r \sim Exp(1)$ and $\alpha$ is given by equation \ref{log_acceptance_ratio_aa}
  \begin{equation}
    \phi_g = 
    \begin{cases}
      \Delta M_a', \Delta \eta_a',	& \text{if } -r < \alpha \\
      \Delta M_a, \Delta \eta_a,	& \text{else}  \\
    \end{cases}
     \label{a/r_aa}
  \end{equation}  
  \begin{equation}
    \alpha = P_a(\Delta M_a', \Delta \eta_a'|\vec{\zeta}, \vec{\phi_z}) - P_a(\Delta M_a, \Delta \eta_a|\vec{\zeta}, \vec{\phi_z})
    \label{log_acceptance_ratio_aa}
  \end{equation}
  

  
  \section{FONSE}
  
\end{document}